lara: merge common functions

lahm86 opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This merges some of Lara's common functions into TRX, but also moves previous Item functions that were used exclusively by Lara  into her module as well for clarity. `Lara_MovePosition` will allow us to offer the walk to items feature in TR2 later.

I'd appreciate some checks for pickups and switches in particular (these call `Lara_MovePosition`), plus in TR1, with and without walk to items enabled. Other checks are quite general, such as Lara responding to geometry if for example mid-roll, or being nudged by walls while swimming etc.

Also, I'm happy to delay merging this until after 4.10 is out.

lahm86 committed:
lara: port Lara_UpdateRoom to TRX

This converts Item_UpdateRoom to Lara_UpdateRoom as its only use case
is as such, and moves it to TRX.

d

d

d

lahm86 committed:
lara: port Lara_ShiftCol to TRX

This converts Item_ShiftCol to Lara_ShiftCol as its only use case
is as such, and moves it to TRX.

lahm86 committed:
items: port Item_GetDistance to TRX

lahm86 committed:
lara: port Lara_IsNearItem to TRX

lahm86 committed:
lara: port Lara_MovePosition to TRX

lahm86 commented:
Just realised I was a few commits behind develop, updated now.

walkawayy reviewed (approved):
Everything LGTM in TR1 and 2. I tried TR1 with animated interactions on and off.

rr- commented on src/libtrx/game/lara/common.c:
```diff
@@ -243,3 +248,174 @@ void Lara_AlignPosition(const ITEM *const item, const XYZ_32 *const vec)
+    lara_item->floor = Room_GetHeight(sector, x, y, z);
+    if (lara_item->room_num != room_num) {
+        const int16_t item_num = Item_GetIndex(lara_item);
+        Item_NewRoom(item_num, room_num);
```
Can we rename (in a separate commit) this function to `Item_UpdateRoom`? Thanks.

lahm86 replied:
Separate commit added with this. I also removed the `if (item->room_num != room_num)` we had in so many places before calling this into the function itself as IMO it makes sense to do the check there rather than repeated throughout.

rr- replied:
Agreed – thank you.

rr- commented on src/libtrx/game/lara/common.c:
```diff
@@ -243,3 +248,174 @@ void Lara_AlignPosition(const ITEM *const item, const XYZ_32 *const vec)
+    coll->shift.z = 0;
+    coll->shift.y = 0;
+    coll->shift.x = 0;
+}
```
Let's try to clean up her APIs a bit and move things that are used mostly inside the lara namespace to `misc.c` or something similar, and here leave only basic functions that are used publicly. Maybe `col_utils.c` for the utility functions needed by `col.c`.
I haven't checked the usages of this function `Lara_ShiftCol` – it's more of a general comment.

lahm86 replied:
Thanks, I wasn't too sure about the line between common and misc, but this makes sense. Will update later.

lahm86 replied:
`Lara_UpdateRoom` and `Lara_ShiftCol` are now in `misc.c` as they're only used by other Lara modules. We can perhaps see how thing pan out the more we move to TRX and see about potentially introducing other modules too.

rr- commented on src/libtrx/game/lara/common.c:
```diff
@@ -243,3 +248,174 @@ void Lara_AlignPosition(const ITEM *const item, const XYZ_32 *const vec)
+        } else {                                                               \
+            source = target;                                                   \
+        }                                                                      \
+    } while (0)
```
Let's `#undef` this macro after we're done with the usage.
Or on the contrary – make this macro public. I think it's worth checking if we have more usages of this pattern (not the macro) across the code base – I think we have quite a lot.

lahm86 replied:
Made this public in `items/utils.h`.

rr- commented on src/tr2/decomp/decomp.c:
```diff
@@ -77,54 +77,6 @@ void CutscenePlayer1_Initialise(const int16_t item_num)
-    );
-    // clang-format on
-}
-
```
This is such a bad function name lol. Thank you for clearing this up.

rr- reviewed (commented):
LGTM. A couple of cosmetic remarks.

lahm86 committed:
items: rename Item_NewRoom

This renames Item_NewRoom to Item_UpdateRoom and also moves the
repeated checks prior to calling this function throughout the codebase,
where the room number is compared with the item's current room number,
to the function itself.

lahm86 commented:
@aredfan Would you mind doing a quick check again as the last commit touches on a few things? Just need to ensure items move between rooms OK mainly, a sanity check would be appreciated.

aredfan reviewed (approved):
> @aredfan Would you mind doing a quick check again as the last commit touches on a few things? Just need to ensure items move between rooms OK mainly, a sanity check would be appreciated.

No problem, I don't mind. I did another test for both games and this PR still looks good. 👍

rr- commented on src/libtrx/game/lara/common.c:
```diff
@@ -1,11 +1,16 @@
 
+#define MOVE_ANIM_VELOCITY 12
+#define MOVE_SPEED 16
+#define MOVE_ANGLE (2 * DEG_1) // = 364
```
Nitpick, I'm trying to low-key introduce a convention for local defines that uses `M_…` syntax for their naming.

lahm86 replied:
Ok, thanks. Will bear in mind going forward too.

rr- commented on src/libtrx/game/objects/general/door.c:
```diff
@@ -188,7 +188,7 @@ static void M_Initialise(const int16_t item_num)
 
         const int16_t prev_room = item->room_num;
-        Item_NewRoom(item_num, room_num);
+        Item_UpdateRoom(item_num, room_num);
```
I'm a bit worried about the early exit here if the rooms match. Doesn't this function rely on the side effects to happen even if the room matches?

lahm86 replied:
`Item_Initialise` does this:

```c
ROOM *const room = Room_Get(item->room_num);
item->next_item = room->item_num;
room->item_num = item_num;
```

So an item is always linked to its room. The side effects of `Item_UpdateRoom` are purely to move these item links around; if the room number is the same, that work is unnecessary as it's already been done on initialisation, or by previous `Item_UpdateRoom` calls where the numbers have differed.

rr- replied:
Just making sure – so there's no way this is initialized wrong, and the side effects of Item_UpdateRoom fix the links that were wrong, right? And the newly introduced `if` inside is safe to rely on data in the `item` in all of those cases.

I'm speaking from my memory but I remember some crashes when working with door.c that were easy to trigger due to some optimistic assumptions about its state. I'm just not sure if it's that. Just a double check kind of thing.

lahm86 replied:
Thanks. I've done some extra checks around doors, bacon Lara and the skidoo and it checks out. Interestingly, with Bacon Lara, prior to this, the code was always removing her from the current room, then re-adding her to the same room, so this is an improvement there. I've tested this extensively on a custom level too with different rooms, and it behaves as expected.

rr- commented on src/libtrx/game/lara/misc.c:
```diff
@@ -28,3 +28,29 @@ void Lara_GetCollisionInfo(const ITEM *const item, COLL_INFO *const coll)
         LARA_HEIGHT);
 }
+
+void Lara_UpdateRoom(const int32_t height)
```
Sorry about this back and forth but this may be now confused with `Item_UpdateRoom` signature which takes `room_num` as its integer argument, rather than height. Maybe something such as `Lara_UpdateRoomToHeight`, or `Lara_UpdateRoomWithOffset`…

lahm86 replied:
No problem, updated.

rr- commented on src/tr1/game/objects/creatures/bacon_lara.c:
```diff
@@ -80,7 +80,7 @@ static void M_Control(const int16_t item_num)
         item->rot.y = g_LaraItem->rot.y - DEG_180;
         item->rot.z = g_LaraItem->rot.z;
-        Item_NewRoom(item_num, g_LaraItem->room_num);
+        Item_UpdateRoom(item_num, g_LaraItem->room_num);
```
Same comment as door.c about forced side effects
(Pickups have that too but I don't think there's anything special about those)

lahm86 replied:
Same as above, and for pickups I think when they're carried they're assigned `NO_ROOM` so the updating will always kick in in that case.

rr- commented on src/tr2/decomp/skidoo.c:
```diff
@@ -874,20 +874,16 @@ int32_t Skidoo_Control(void)
-        Item_NewRoom(g_Lara.item_num, room_num);
-    }
+    Item_UpdateRoom(g_Lara.skidoo, room_num);
+    Item_UpdateRoom(g_Lara.item_num, room_num);
```
Flagging it as a risky update too.

lahm86 replied:
Lara and the skidoo always have the same room number - I don't feel this is risky, unless I'm perhaps missing something?

rr- replied:
I haven't fully followed the code, only saw there was no `if` and wanted us to pay extra attention to it if `room_num` of both items can be relied on in this case.

lahm86 committed:
lara: rename Lara_UpdateRoom

rr- reviewed (approved):
Thank you.
